src/gridfs.rs

```diff
@@ -0,0 +1,285 @@
+pub mod options;
+
+use core::task::{Context, Poll};
+use std::{io, pin::Pin};
+
+use crate::{
+    concern::{ReadConcern, WriteConcern},
+    cursor::Cursor,
+    error::{Error, Result},
+    selection_criteria::SelectionCriteria,
+    Database,
+};
+
+use options::*;
+
+use bson::{oid::ObjectId, Bson, DateTime, Document};
+use futures_util;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+// Contained in a "chunks" collection for each user file
+struct Chunk {
+    id: ObjectId,
+    files_id: Bson,
+    n: i32,
+    // default size is 255 KiB
+    data: Vec<u8>,
+}
+
+// A collection in which information about stored files is stored. There will be one files
+// collection document per stored file.
+pub struct FilesCollectionDocument {
+    id: Bson,
+    length: i64,
+    chunk_size: i32,
+    upload_date: DateTime,
+    filename: String,
+    metadata: Document,
+}
+
+/// Struct for storing GridFS managed files within a [`Database`].
+pub struct GridFsBucket {
+    // Contains a "chunks" collection
+    pub(crate) bucket_name: String,
+    pub(crate) db: Database,
+    pub(crate) chunk_size_bytes: i32,
+    pub(crate) read_concern: Option<ReadConcern>,
+    pub(crate) write_concern: Option<WriteConcern>,
+    pub(crate) read_preference: Option<SelectionCriteria>,
+}
+
+// TODO: RUST-1399 Add documentation and example code for this struct.
+pub struct GridFsUploadStream {
+    pub id: Bson,
+}
+
+impl GridFsUploadStream {
+    /// Consumes the stream and uploads data in the stream to the server.
+    pub fn finish(self) {
+        todo!()
+    }
+
+    /// Aborts the upload and discards the upload stream.
+    pub fn abort(self) {
+        todo!()
+    }
+}
+
+impl tokio::io::AsyncWrite for GridFsUploadStream {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<tokio::io::Result<usize>> {
+        todo!()
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        todo!()
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<tokio::io::Result<()>> {
+        todo!()
+    }
+}
+
+pub struct GridFsDownloadStream {
+    pub id: Bson,
```

Comment:
rather than exposing this field publicly, I think it might be preferable to use a helper. That way it's impossible for users to mutate it. Ditto for the upload stream.